### DISCUSSION TITLE
Project Update April 2023

### DISCUSSION THREAD

**Discussion by DeeEmm** [DIY-Flow-Bench/DIY-Flow-Bench#140] [Announcements]
Commit https://github.com/DeeEmm/DIY-Flow-Bench/commit/acd61cbf529a911fe4d0c7585ba3941486a491ad contains Build 23040301

```
Build #         - Description of Change

23040301        - Added support for Bosch 0280218067 MAF
                - Added transfer function for MAF housing diameters
                - Removed Siemens 5WK9605 MAF Data
23033101        - Added diameter var to MAFData files
23033001        - Added yml template for Support forum Posting Guidelines
```

Another month, and another updated thread.

The latest commit includes both updated MAF data files and inclusion of the Bosch 0280218067 from the Audi RS4. This is a high flowing MAF that is easier to find in Europe than the LS2 MAF sensors.

The update also includes the transfer function for the MAF Housing Diameter. This compares the standard MAF diameter value from the MAF Data file with the user saved diameter from the config settings. If the two values are different then the flow value is transposed based on the change in housing area. This allows users to use slot style MAF sensors in custom diameter housings, effectively extending the flow capabilities of the MAF.

There is one caveat to this in that there will be a limit in the size change as at some point the efficiency of the sensor will be impacted. I'm not entirely sure what this limit is, but small increases are likely to be fine (maybe < 20% ?) but large increases will no doubt directly impact sensor accuracy.

**Comment by DeeEmm** [comment]
BTW if you are looking for the GUI elements in above commit they are missing but you can find them in commit https://github.com/DeeEmm/DIY-Flow-Bench/commit/7a2b8deaacc0ce50266aa295c98dd9f75399a60c

```
Build #         - Description of Change

23040302        - Added GUI handling for MAF diameter
```

**Comment by DeeEmm** [comment]
I've also started to work though the MAF Data files and identifying those that have validated data. What I want to end up with is both validated and tested data for the MAF data files we include with the project. This will assist project uptake and anyone trying to source MAF sensors. 

I will also create a dedicated support thread for each MAF data file. The first one is done (for the Bosch 0280218067), I'll roll the others out over the next few weeks.

**Comment by DeeEmm** [comment]
I've added the SIEMENS_5WK9605 MAF Data file back with updated and verified data in commit https://github.com/DeeEmm/DIY-Flow-Bench/commit/18de31b1a4346bb35d3098d0589da4f39b42be61

```
Build #         - Description of Change

23040401        - Added updated SIEMENS_5WK9605 MAF Data file
```

I've also created an official discussion thread for the SIEMENS_5WK9605 MAF - https://github.com/DeeEmm/DIY-Flow-Bench/discussions/141

This data is from https://www.ms4x.net/index.php?title=Aftermarket_Upgrade_Sensor_Data

**Comment by black-top** [comment]
I have already received a new pressure and maf sensors. I will start testing as soon as I have some time. The next step will probably be pressure management. I don't remember if we talked about it, but one of the pressure control options could be an electronic throttle. Simple control with h-bridge and has two potentiometers for close loop feedback. It would probably be cheaper and simpler than making a custom valve with a stepper motor.

**Comment by DeeEmm** [comment]
Yes, it was briefly discussed.  My plans are to use the inbuilt DACs to generate analog output reference that can then be used to drive whatever is required (H-Bridge, VFD, etc). These are 0-3v so may need to be level shifted to work with whatever circuit or device they are connected to. There is also the option for stepper motor control as well. Plenty of of options to suit all kinds of solutions.

An e-throttle could be a valid and somewhat elegant solution, but despite making EFI ECU's I've never actually played with them. 

AFAIK the throttle plates are spring loaded to close if the power is removed (fail safe). This could mean that a simple MOSFET and PWM control is all that is required to 'hold' them at any given position. This is how a basic IAC valve works. However they do suffer from hysteresis and far better control is had by stepper based IACVs . 

PWM control  has a tendency to 'hunt', that is it overshoots the target, the power is removed, it overshoots the target in the other direction, the power is reapplied, etc, ad infinitum. This is why good PID control is VERY important for correct modulation with PWM controlled systems, especially those containing springs. 

With stepper based control, the valve is simply sent to the desired position and it stays where it is sent. Heaps more accurate, heaps more simple to control and they do not suffer from mechanical hysteresis.

However ideally in all cases PID control needs to be employed simply as you cannot easily control a closed loop system without it

The only issue I can see is that PID adds a level of complexity that needs to be dialled in for every system. PID constants will not universally apply to all systems even if they use the same components. I would say that PID tuning is out of the skill set of most amateur hackers and it concerns me that it will be difficult for most people to set up.

With all this said, pressure / fan speed control is not something that is planned for the V2 release. 

Initially the system will work with just manual control.

BTW. I would say that the potentiometers are not needed as it is the reference pressure that determines the valve position and provides the feedback

**Reply by black-top** [reply]
it's a good idea to use a single mosfet instead of a full h-bridge. But the question is if the valve would be stable. Closed loop with pressure also better solution, not with tps. I agree , it can be difficult to adjust the PID, especially for the inexperienced. But it would be very cheap solution to control the pressure. 
You are right, the stepper does not have this problem. In order not to need a powerful stepper motor, can use a worm gear reducer that does not rotate due to the air flow, but still PID nessesary to stable controll pressure.  I wonder why oem doesn't use a stepper in dbw

**Reply by DeeEmm** [reply]
> But the question is if the valve would be stable

I just found this video of someone controlling a DBW unit using PWM and it seems to work fine, but this is only on the bench

https://youtu.be/H2KJgYzh3lc

> I wonder why oem doesn't use a stepper in dbw

I think some Toyota models use steppers

**Reply by black-top** [reply]
Yes, but here is working on constant spring pressure. The airflow will create varying resistance to the motor depending on the angle of the valve.

**Reply by DeeEmm** [reply]
For the fail-safe to work the spring pressure must be high enough to overcome the engine vacuum / air flow so I would think that it would be fine. There is a very high gear ratio so I suspect that the spring pressure is quite high.

It's a similar story with a regular cable operated throttle body. The cable only operates in one direction to open the throttle plate. It is the spring that returns it closed.

The throttle plate is also symmetrical so it theoretically has similar forces acting on both sides of the shaft

Of course there is only one way to know for sure which is why I bought one to test with

**Reply by black-top** [reply]
Probably yes, I didn't think about it. Single mosfet control greatly simplifies the entire system control.

**Comment by DeeEmm** [comment]
In the interests of research I just bought a cheap pre-loved DBW throttle body. This one is from a VZ Commodore. Bosch Part No. 12589056. It's also used on Alfa 159 and maybe some other European cars.

![s-l1600](https://user-images.githubusercontent.com/3038710/230236230-853ad0ef-8b01-4507-bfa3-ffbfc5d97e9c.jpg)

I suspect that all Bosch units will function in a similar manner so any code that works for this should work for other 12V DBW TBs

These are super common here in Aus and dirt cheap. I only paid AU$30 (about 15 Euros).

Another project to add to the increasingly long list of things to do. Hopefully I'll live to be 160, then I might have a chance of getting them all done lol

**Reply by black-top** [reply]
Most bosch DBW throttles are very similar and they are used in almost all European cars. Internally they are all identical. My programming skills are very limited, so I can only help with testing and ideas.

**Reply by DeeEmm** [reply]
The code side for PWM is actually very simple. 

You can also use DAC's to generate a simple analog output reference voltage and then use an external stand alone PWM controller, which is even simpler.

These take an analog reference and output PWM at whatever voltage you need. They are very cheap to buy and are used in various Arduino control projects. I've used them before for laser module and DC spindle motor control. I suspect that this is what is being used in the video above.

This is what I've used in the past. These are 300W which would easily drive the TB

![s-l500](https://user-images.githubusercontent.com/3038710/230356463-927a5f2d-02c8-4303-83da-fa21c43bf97b.jpg)

Then there are these...

![s-l1600 (1)](https://user-images.githubusercontent.com/3038710/230357524-79a760d3-1e4d-46e1-92e1-6b97027bd6b3.jpg)

Which I think are what he's using in the video. These are 3kw, which I think is overkill

As you can see these will also take a basic potentiometer input as well as an analog reference

**Reply by DeeEmm** [reply]
Just checked. I have one of the 300W PWM units spare, so I can easily do a quick test using the potentiometer when the TB arrives.

**Reply by black-top** [reply]
will be interesting to hear the results!

**Comment by DeeEmm** [comment]
Commit https://github.com/DeeEmm/DIY-Flow-Bench/commit/fdc594027036da3e6636f3fc038ee9f643e655c2 contained Build 23040901 which includes the following changes


```
Build #         - Description of Change

23040901        - Fixed error with MPXV7025 info display
                - Fixed error with MPXV7025 transfer function
                - Fixed error with MAF Data scaling
23040601        - Added lift data handling
                - liftdata.json file management in spiffs
                - liftdata.json export via WebGUI
23040402        - Updated GUI Graph data handling
23040401        - Added updated SIEMENS_5WK9605 MAF Data file
```

**Comment by DeeEmm** [comment]
Commit contains Build 23041101 which incorporates the following changes

```
Build #         - Description of Change

23041101        - Basic lift data data handling finalised
23041001        - increased loop task size in platformio
```

It is advised to update your platformio file as the loop task stack size has been increased. If you do not do this then you will most likely see some stack overflow errors. This is because the loop task data has grown in size and now exceeds the standard allocated size.

The other change is the finalisation of the lift graph data handling. This functions as follows.

## Lift data recording
To record a lift data point, click the 'Save Data' button on the Dashboard to show the lift capture dialog. Select the lift value that you are measuring and then click the 'Capture' button to save the current flow data. 

The data is saved to the liftdata.json file in SPIFFS memory and also displayed on the lift graph.

**NOTE**: Lift point values are determined by the 'Valve Lift Interval' setting on the GUI config tab.

## Lift Graph
Clicking on the 'Data Tab' displays the Lift Graph. This is a graphical representation of the lift data stored in liftdata.json

Currently this data is very basic, but it gives you an opportunity to visualise the flow characteristics of the object you are measuring. 

Currently there is no way to natively export an image of the Lift Graph but you can take screen captures using 'screenshots' or whatever your device or operating system supports to manually compare to other traces. 

## Lift Data Export
To save the Lift Data to your local computer, click the 'Export' button. This will open a save dialog that will allow you to save the liftdata.json file to your local device.

The data is in JSON format which can easily be imported into Excel or other programs

You can also access the liftdata.json file via the 'Configuration > Files' dialog.

---

I wanted to bring this functionality into to the V2 release to round out the basic feature set as including a method for saving data makes the project a much more usable tool. Even with just basic view and export functions as included.

Whilst the data is very basic I do have a working example of hover data values, where hovering over a stored datapoint displays the actual data value. I also need to tie the graph scaling to the MAF max flow value as currently the scale values are fixed. These two features will be added before the final V2 release.

Future plans for the Data graph are largely tied to the SD card support as to be able to load multiple graphs for comparison there needs to be an alternative method of data storage to the SPIFFS file system. The reason for this is that the SPIFFS file storage is size limited. There's also lots of opportunity to make improvements to the way that the data is presented as HTML5 and Javascript provide almost unlimited possibilities. Further changes will however will have to wait until V3

It is likely that the V2 release will remain current for several months, so for me it is important to present a solution that is fully functional. Whilst I am trying to ensure that we have frequent interim updates published, V3 will be several months away so ensuring that the V2 release is able to stand on its own feet in the interim is important for project acceptance and reputation.

**Comment by DeeEmm** [comment]
Commit https://github.com/DeeEmm/DIY-Flow-Bench/commit/0bbc4ed25fcf938128a73b0c19fba3ba952928d7 contains Build 23041201 which includes the following changes

```
Build #         - Description of Change

23041201        - Updated MAf conversion code
                - Implimented simplified Mass flow to CFM conversion
                - Updated ACDELCO_92281162 MAF Data
23041102        - Changed 'Save Data' button test to 'Capture'
```

**Comment by DeeEmm** [comment]
Commit https://github.com/DeeEmm/DIY-Flow-Bench/commit/4cdba3a50a699f610a66c3014eb8b8fc18ba8d14 contains Build 23041202 which incorporates the following changes

```
Build #         - Description of Change

23041202        - Added adjustment for humidity and ref pressure to CFM Calc
```

**Comment by DeeEmm** [comment]
Commit https://github.com/DeeEmm/DIY-Flow-Bench/commit/07e7ef3a1329150be0e15eb63fc2e11218b3ff7a contains Build 23041301 which comprises of the following changes

```
Build #         - Description of Change

23041301        - Added graph image capture
                - General code tidy
```

This marks a bit of a milestone for the project in that all (planned and unplanned) features for V2 are now finished and working (even those additional features that were intended for V3). I'm a lot happier with the current featureset as the additional features recently added make the project much more usable. It feels a lot more complete now.

MAF Data handling is now working to an acceptable level for several MAF sensors and the GUI data handling now provides the user with basic tools to view data within the browser in a graphical format as well as export the data to the users local device both in JSON and PNG format. These function really leverage the platforms ability to help streamline using the bench to record and manage data. There's no need to manually note anything down or manually draw lift plots, everything is taken care of in the browser and can be exported as usable data. This is the actual data graph image exported from the browser. 

![LiftGraph](https://user-images.githubusercontent.com/3038710/231637176-68ebe779-ebe5-40fe-8f5e-769b8bbb7c86.png)

(I know it looks a little fuzzy, but the rasterisation of the SVG data by the browser is a little limited. )

There's a lot more than can be done data-wise, but to avoid any more scope creep, all features are now locked for the V2 release. 

This is the closest that the project has been to the final V2 release version and I'm excited that the finish for this version is finally in sight. I suspect that there are still going to be some bugs to address as we are still very much in the middle of validation testing, but I'm going to roll this build up into a new release (RC6). If you are running any version of the code I recommend that you update and test drive the new code. Most importantly if you find any issues with it, please let us know. 

In addition to the code update I also have the new shield assembled. I'm working through testing it but so far everything looks good.

![PXL_20230413_025536457 MP](https://user-images.githubusercontent.com/3038710/231636672-d0298b9f-226d-4a58-b848-9705e597321d.jpg)

I've got a few spare PCBs which I'm going to send out to the early adopters and contributors once I've updated the documentation. The shield too is very close to being finalised. Fingers crossed I don't find any more bugs and this version can be released. 

Happy days :)

**Comment by DeeEmm** [comment]
Commit [fe90cd9](https://github.com/DeeEmm/DIY-Flow-Bench/commit/fe90cd9c53009b9ea783d8a526bfb258ea4f9958) contains Build 23042301 which incorporates the following changes

```
Build #         - Description of Change

23042301        - Updated convertFlowDepression() method 
23041401        - Updated package.json...
                - Moved dependencies into devDependencies section and made private
```

The package.json dependencies are not directly related to the project but are included to help people set up their environment who want to use gulp to minify the html, javascript and css files. I've changed the way that these are presented within package.json to hopefully stop them being flagged.

The update to the convertFlowDepression() method is to detect whether the target reference pressure is higher or lower than the source reference pressure. Previously it always assumed that the target was higher.
